### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by janeliu-slac**
Adding unit tests.

### CODE REVIEW

File: lcls_naming_tool/__init__.py
Comment (tangkong) [inline lcls_naming_tool/__init__.py:1]: You actually need this file for the python files in the directory to be gathered for import.  
https://docs.python.org/3/tutorial/modules.html#packages

File: lcls_naming_tool/tests/test_cli.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def test_display_version(capsys):
+
+    display_version()
+    captured = capsys.readouterr()
+    assert str(captured.out).strip() == 'v1.0.0'
+
+
+load_taxons()
```
Comment (tangkong) [inline lcls_naming_tool/tests/test_cli.py:17]: Perhaps I missed this in the previous review, but what is the scope of this function call.  We just need to call it once per session and the taxons are loaded into the module's global scope?

File: lcls_naming_tool/tests/test_cli.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+@pytest.fixture
+def invalid_name():
+    return ['ABC12345M9', 'ABC', 'NBC', '7', '_PCTRLSPRBCK']
+
+
+def test_display_version(capsys):
+
+    display_version()
+    captured = capsys.readouterr()
+    assert str(captured.out).strip() == 'v1.0.0'
```
Comment (tangkong) [inline lcls_naming_tool/tests/test_cli.py]: This test will start to fail once we update the version number

File: lcls_naming_tool/tests/test_cli.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+
+load_taxons()
+
+
+def test_functional_component_is_valid_true(valid_name):
```
Comment (tangkong) [inline lcls_naming_tool/tests/test_cli.py]: Pytest lets you parametrize these tests so you don't have to copy paste.  This also lets you expand the test matrix more easily as we find and want to test more edge cases.  See: https://docs.pytest.org/en/stable/how-to/parametrize.html
```python
VALID_NAMES = ['MR2K4', 'KBO', 'PIP', '01', 'PUMPSIZE']


@pytest.mark.parametrize('name', VALID_NAMES)
def test_valid_name(name: str):
    assert functional_component_is_valid(name)
    assert fungible_is_valid(name)
```

And similarly for invalid names etc

File: lcls_naming_tool/tests/test_cli.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+# Unit tests for the LCLS Naming Tool script
+
+import pytest
+import json
+from pathlib import Path
+from lcls_naming_tool import display_version, load_taxons, functional_component_is_valid, fungible_is_valid, constituent_component_is_valid, increment_is_valid, starts_alphanumeric, validate
+
+
+@pytest.fixture
```
Comment (tangkong) [inline lcls_naming_tool/tests/test_cli.py]: This is marked as a fixture but never called like a fixture.  Pytest auto-fills argument names that match fixture names, so a test defined as below would be valid:

```python
def test_thing(valid_name: list[str]):
    ...
```

File: lcls_naming_tool/tests/test_cli.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+def valid_name():
+    return ['MR2K4', 'KBO', 'PIP', '01', 'PUMPSIZE']
```
Comment (tangkong) [inline lcls_naming_tool/tests/test_cli.py]: It appears there are functional/fungible components mixed in here. I would argue that these should be separate fixtures, so we can test the cases more completely

**Review by tangkong — COMMENTED** [review]
This is a good start.  I think I'll want to see a much more exhaustive list of test cases in the near future, and I've suggested some ways of making it easier to do that in pytest.

File: lcls_naming_tool/__init__.py
Comment (janeliu-slac) [reply lcls_naming_tool/__init__.py:1]: Hmm strangely, `pytest -v` doesn't seem to work when `__init__.py` is in that folder. Is there a way to get around this?

Comment (tangkong) [reply lcls_naming_tool/__init__.py:1]: ~Is there an error message to refer to?~

Because this package is not installed, python doesn't recognize `lcls_naming_tool` as an importable package.  You can do a couple things, only the first of which I think is proper:
* set this up to installable, and install it into your dev environment to test (from the repo root: `pip install -e .`)
* change the import to `from ..lcls_naming_tool`.  This searches the parent directory for a python file named "lcls_naming_tool", ignoring the installed packages
  * This will probably get your tests to run, but we'll want to revert this once we make the package installable

Comment (tangkong) [reply lcls_naming_tool/__init__.py:1]: It looks like most of the cookiecutter needed to install this package is in place.  Activate your venv and give `pip install -e .` a try (the `-e` means editable, letting your modify the source files and see the result in your imports without reinstallation)

Comment (janeliu-slac) [reply lcls_naming_tool/__init__.py:1]: Ah, that's what I couldn't figure out. Embarassed to admit this is my first time creating an installable python package. 😅 Got it working now.

Comment (tangkong) [reply lcls_naming_tool/__init__.py:1]: Nothing to be embarassed about, everyone has to do everything for the first time at least once

File: lcls_naming_tool/tests/test_cli.py (hunk 1 above)
Comment (janeliu-slac) [reply lcls_naming_tool/tests/test_cli.py:17]: Yep, I believe this is what is happening. `load_taxons()` modifies global variables declared in lcls_naming_tools.py. The global variables belong to the scope of the `lcls_naming_tool` module.

File: lcls_naming_tool/tests/test_cli.py (hunk 2 above)
Comment (janeliu-slac) [reply lcls_naming_tool/tests/test_cli.py]: Updated to use `os.popen()` to get output from `git describe --tags`.

File: lcls_naming_tool/tests/test_cli.py (hunk 3 above)
Comment (janeliu-slac) [reply lcls_naming_tool/tests/test_cli.py]: Hmm when I do this the `assert fungible_is_valid(name)` seems to come back with an error. Oh, I see it's testing each of the names in the assert statements.
```
E       AssertionError: assert False == True
E        +  where False = fungible_is_valid('MR2K4')
```

Comment (tangkong) [reply lcls_naming_tool/tests/test_cli.py]: If you hit an `AssertionError`, that means your test is failing!

Comment (janeliu-slac) [reply lcls_naming_tool/tests/test_cli.py]: Ok, think I figured out how @pytest.mark.parametrize works. Updated the tests.

Comment (tangkong) [reply lcls_naming_tool/tests/test_cli.py]: If in doubt, you can always look through `pcdshub` for examples.  We use pytest rather extensively

**Comment by janeliu-slac** [comment]
I'm going to look through the pytests in pcdshub tomorrow and probably study up on it this weekend. I think pretty much all of the issues are fixed (will add a more exhaustive list of test cases in the next PR). If all looks good could it be merged? (I think I'm not authorized to merge)

File: lcls_naming_tool/tests/test_cli.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def test_display_version(capsys):
+    display_version()
+    captured = capsys.readouterr()
+    curr_ver = os.popen('git describe --tags').read().strip().split('-')
+    assert str(captured.out).strip() == curr_ver[0]
+
+
+load_taxons()
```
Comment (ZLLentz) [inline lcls_naming_tool/tests/test_cli.py:17]: Small optimization for later: there's probably a better time and place to load the taxons. Perhaps this should be done automatically when checking anything if it hasn't been done yet.

**Review by ZLLentz — APPROVED** [review]
I don't see any compelling reason not to merge this.

Maybe I'll add the standard github actions stuff so the unit test results will be displayed here in follow-up PRs.

**Review by tangkong — APPROVED** [review]
I would like to see more test cases (read: more entries in the paramterization), but that will grow as we develop the tool.  This looks good to me for now 👍

File: lcls_naming_tool/tests/test_cli.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
+# Unit tests for the LCLS Naming Tool script
+
+import pytest
+import json
+import os
+from pathlib import Path
+from lcls_naming_tool.lcls_naming_tool import display_version, load_taxons, functional_component_is_valid, fungible_is_valid, constituent_component_is_valid, increment_is_valid, starts_alphanumeric, validate
+
+
+def test_display_version(capsys):
```
Comment (ZLLentz) [inline lcls_naming_tool/tests/test_cli.py:10]: Something to work on later: this test does not pass

Comment (janeliu-slac) [reply lcls_naming_tool/tests/test_cli.py:10]: Hmm, is it the method of testing that doesn't pass or is pytest returning an error? When I run `pytest -v` I get a `test_cli.py::test_display_version PASSED` message.

Comment (tangkong) [reply lcls_naming_tool/tests/test_cli.py:10]: you can see the test logs under the actions tab of this repo 
https://github.com/pcdshub/lcls_naming_tool/actions/runs/11075882558/job/30778257851#step:17:85

looks like argparse is pulling args from the pytest invocation.  I think that if you want to use `sys.args` in pytest you need to scrub the old args and provide new ones
